File: mcp/src/tools/ddgs_search.py
# pylint: disable=line-too-long
"""DuckDuckGo Search tool."""
import json
import sys

try:
    from duckduckgo_search import DDGS
except ImportError:
    print(json.dumps({"error": "duckduckgo_search package not found. Please install it using 'pip install duckduckgo-search' (preferably in a virtual environment)."}), file=sys.stderr)
    sys.exit(1)

if len(sys.argv) != 3:
    print(json.dumps({"error": f"Invalid argument count. Expected 2 (query, max_results), got {len(sys.argv) - 1}."}), file=sys.stderr)
    sys.exit(1)

try:
    query = sys.argv[1]
    max_results = int(sys.argv[2])
    results = DDGS().text(query, max_results=max_results)
    print(json.dumps(list(results)))
except Exception as e:
    print(json.dumps({"error": str(e)}), file=sys.stderr)
    sys.exit(1)

File: mcp/src/tools/ddgs.search.ts
import { z } from "zod";
import { runCommand } from "../lib/shell.js";

export const DdgsSearchInputSchema = z.object({
  query: z.string(),
  maxResults: z.number().optional().default(5),
});

import path from "path";
import { fileURLToPath } from "url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export async function ddgsSearchHandler(args: z.infer<typeof DdgsSearchInputSchema>) {
  const scriptPath = path.join(__dirname, "ddgs_search.py");

  // Sanitize query by removing control characters, newlines, and null bytes
  // eslint-disable-next-line no-control-regex
  const safeQuery = args.query.replace(/[\x00-\x1F\x7F-\x9F]/g, " ").trim();
  if (!safeQuery) {
    throw new Error("Query cannot be empty after sanitization.");
  }

  const result = await runCommand("python3", [scriptPath, safeQuery, (args.maxResults ?? 5).toString()]);

  // duckduckgo_search logs annoying deprecation warnings to stderr we want to ignore
  // if it really failed, exitCode will be non-zero and we'll have stdout/stderr
  if (result.exitCode !== 0) {
    let parsedError = result.stderr || result.stdout;

    const lines = parsedError.split("\n");
    for (const line of lines) {
      try {
        const json = JSON.parse(line);
        if (json.error) {
          parsedError = json.error;
          break;
        }
      } catch {
        // continue
      }
    }

    throw new Error(`Failed to search ddgs: ${parsedError}`);
  }

  try {
    const data = JSON.parse(result.stdout);
    if (!Array.isArray(data)) {
      throw new Error("Expected an array of search results");
    }
    return { results: data };
  } catch (e) {
    const message = e instanceof Error ? e.message : String(e);
    throw new Error(`Failed to parse ddgs search results. Error: ${message}. Output was: ${result.stdout}`);
  }
}
